### PR TITLE
Add cached version of os.path.isfile to avoid repetitive I/O

### PR DESCRIPTION
**Pull request by correctmost**
## Type of Changes
|     | Type                   |
| --- | ---------------------- |
| ✓   | :hammer: Refactoring   |

## Description
When running Pylint on codebases like yt-dlp, I see a lot of time spent on repeated file checks:

```
   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
   513208    2.615    0.000    2.640    0.000 {built-in method posix.stat}
   372833    0.309    0.000    2.115    0.000 <frozen genericpath>:27(isfile)
```

This PR introduces a cached version of `isfile` to reduce the cost of these repeated checks.

My thought was that similar cached versions could be introduced for `os.path.exists` and `_io.open` to alleviate issues like https://github.com/pylint-dev/pylint/issues/9603.  Some concerns were raised about caching in the past (#2429), so I'm not sure if this is an acceptable approach for the project.

Nonetheless, I thought I'd submit a PR to try to figure out how to make disk I/O less expensive while running Pylint.

### Before

```
   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
   513208    2.615    0.000    2.640    0.000 {built-in method posix.stat}
   372833    0.309    0.000    2.115    0.000 <frozen genericpath>:27(isfile)
```

| Command | Mean [s] | Min [s] | Max [s] | Relative |
|:---|---:|---:|---:|---:|
| `pylint --recursive=y .` | 36.214 ± 0.381 | 35.590 | 36.608 | 1.00 |

### After

```
   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
   180887    1.061    0.000    1.086    0.000 {built-in method posix.stat}
    40526    0.037    0.000    0.255    0.000 <frozen genericpath>:27(isfile)
```

| Command | Mean [s] | Min [s] | Max [s] | Relative |
|:---|---:|---:|---:|---:|
| `pylint --recursive=y .` | 33.968 ± 0.236 | 33.685 | 34.283 | 1.00 |

### Reproduction notes
- The tables were generated by hyperfine
  - `hyperfine --ignore-failure --warmup 2 --runs 5 --export-markdown=baseline.md 'pylint --recursive=y .'`

<details>
<summary>cProfile script</summary>

```bash
git clone https://github.com/yt-dlp/yt-dlp.git
cd yt-dlp
git checkout ef36d517f9b05785d61abca7691d9ab7d63cc75c

cat << EOF > ./profile_pylint.py
import cProfile
import pstats
import sys

sys.argv = ['pylint', '--recursive=y', '.']
cProfile.run('from pylint import __main__', filename='stats')

with open('profiler_stats', 'w', encoding='utf-8') as file:
    stats = pstats.Stats('stats', stream=file)
    stats.sort_stats('tottime')
    stats.print_stats()
EOF

cat << EOF > .pylintrc
[MAIN]
jobs=1

[MESSAGES CONTROL]
disable=all
enable=E0401

[REPORTS]
reports=no
score=no
EOF

python ./profile_pylint.py

```
</details>

## Related I/O issues
- https://github.com/pylint-dev/pylint/issues/9310
- https://github.com/pylint-dev/pylint/issues/9603
- https://github.com/pylint-dev/pylint/issues/9613

### CODE REVIEW

File: astroid/modutils.py [hunk 1]
```diff
@@ -604,6 +604,12 @@ def is_relative(modname: str, from_file: str) -> bool:
     )
 
 
+@lru_cache(maxsize=1024)
+def isfile(path: str | os.PathLike) -> bool:
```
Comment (DanielNoord) [inline astroid/modutils.py:608]: Shouldn't it be 
```suggestion
def is_file(path: str | os.PathLike) -> bool:
```

To keep it in line with other utilities?

Comment (Pierre-Sassoulas) [reply astroid/modutils.py:608]: Maybe even
```suggestion
def cached_is_file(path: str | os.PathLike) -> bool:
```

**Review by Pierre-Sassoulas — APPROVED** [review]

**Review by DanielNoord — COMMENTED** [review]
Should this cache be added to our cache clearing mechanism?

**Review by DanielNoord — APPROVED** [review]
Didn't have a good look it seems. Thanks!

**Pull request merged**